frontend/packages/sdk/src/contracts/PinkBlueprint.ts

```diff
@@ -304,4 +322,62 @@ export class PinkBlueprintPromise {
       send: this._decorateMethod((origin: string | AccountId | Uint8Array) => from(inkQueryInternal(origin))),
     }
   }
+
+  async #send(constructorOrId: string, options: PinkBlueprintSendOptions, ...args: unknown[]) {
+    const { cert: userCert, ...rest } = options
+    const txOptions: PinkBlueprintOptions = {
+      gasLimit: options.gasLimit,
+      value: options.value,
+      storageDepositLimit: options.storageDepositLimit,
+    }
+
+    const tx = this.#tx[constructorOrId]
+    if (!tx) {
+      throw new Error(`Constructor not found: ${constructorOrId}`)
+    }
+
+    const address = 'signer' in rest ? rest.address : rest.pair.address
+    const cert = userCert || (await this.phatRegistry.getAnonymousCert())
+    const estimate = this.#query[constructorOrId]
+    if (!estimate) {
+      throw new Error(`Constructor not found: ${constructorOrId}`)
+    }
+
+    const gasPrice = this.phatRegistry.clusterInfo?.gasPrice
+    const depositPerByte = this.phatRegistry.clusterInfo?.depositPerByte
+    if (!gasPrice || !depositPerByte) {
+      throw new Error('No Gas Price or deposit Per Byte from cluster info.')
+    }
+
+    const [clusterBalance, onchainBalance, { gasRequired }] = await Promise.all([
+      this.phatRegistry.getClusterBalance(address),
+      this.api.query.system.account<FrameSystemAccountInfo>(address),
+      estimate(cert.address, { cert }, ...args),
+    ])
+
+    const msg = this.abi.findConstructor(constructorOrId).toU8a(args)
+    const storageDepositRequired = depositPerByte.mul(new BN(msg.length * 2.2))
```

Comment:
But the storage deposit is not related to the size of args?

Reply:
What's the proper way to estimate the storage deposit for contract upload and instantiate?

Reply:
> What's the proper way to estimate the storage deposit for contract upload and instantiate?

They are two steps, so it should be `estimated_deposit_for_code_storage + estimated_deposit_for_instantiation` where:
```
estimated_deposit_for_code_storage = code_size * 2.2;
estimated_deposit_for_instantiation = storage_deposit_limit if charges or 0.
```

Reply:
> But the storage deposit is not related to the size of args?

For example, when instantiating ActionOffchainRollup with JS code, it incurs a high storage deposit fee and requires consideration of the size of args.